docs/debugging/thundershark.md

```diff
@@ -0,0 +1,82 @@
+
+In particular it aids: 
+- profiling - surveying call duration and frame sizes, 
+- debugging - inspecting the parameters and return values passed, 
+- flow analysis - investigating the order of calls and notifications, keeping track of interfaces instance's lifetime. 
+
+It can also have an educational value - by showing their internals dissected it lets developers examine how COM frames are built.
+
+*ThunderShark* consists of a dissector plugin for [*Wireshark*](https://www.wireshark.org) written in Lua and an extension to the *ProxyStubGenerator* (called "LuaGenerator") that compiles the 
+interface definitions to a format understandable by the plugin. *ThunderShark* is thus completely platform and operating system agnostic.
+
+> *Wireshark* version 4.0 or later is required.
+
+## Prerequisites
+
+Firstly, the file ```protocol-thunder-comrpc.lua``` file needs to be placed in *Wireshark*'s plugins folder. In Windows this is typically ```%APPDATA%\Wireshark\plugins```
+or ```%APPDATA%\Roaming\Wireshark\plugins``` folder, while on Linux it's the ```~/.local/lib/wireshark/plugins``` folder.
+
+Secondly, using the *LuaGanerator* tool, interface definitions need to be created. 
+
+Typical usage:
+```
+./GenerateLua.sh [<Thunder_dir> <ThunderInterfaces_dir>]
+```
+```
+LuaGenerator.bat [<Thunder_dir> <ThunderInterfaces_dir>]
+```
+
+The tool will produce a ```protocol-thunder-comrpc.data``` file that holds all interface definitions combined and converted to Lua data tables. This file must be
+placed in the same folder as ```protocol-thunder-comrpc.lua``` file (i.e. *Wireshark*'s plugins folder). This step should be repeated each time there is a change in the interfaces.
+
+> No instrumentation (code changes) of *Thunder* core or plugins is required!
+
+## Configuration
+
+COM servers that are going to be eavesdropped need to be configured to a socket port that can be easily captured, like TCP/IP. This is typically done by adjusting the appropriate
+configuration field.
+
+For example in *Thunder*'s ```config.json```:
+```json
+"communciator":"127.0.0.1:62000",
+```
+
+> Refer to plugin documentation on how to configure custom COM servers provided by plugins (e.g. *OpenCDM*).
+
+This port configuration must be reflected in *ThunderShark*'s config in *Wireshark*: ```Edit/Preferences/Protocols/Thunder COM-RPC Protocol```.
+
+An option to set instance ID size is also provided. It needs to be ensured that it matches the implementation (most often it will be 32-bit long).
+
+## Capture
+
+The COM traffic needs to be captured in a *Wireshark*-readable format, like *pcap*. 
+
+On Linux *tcpdump* is the typical choice, but any other tool able to capture TCP/IP traffic and save it to a *pcap* file can be used. On Windows *Wireshark* itself can be used for this purpose.
+
+Example usage (executed on the DUT, normally before starting Thunder):

```

Comment:
Oops I guess you will probably sya: Basic Knowledge but what is DUT ?

Reply:
Device Under Test :-)

Reply:
Correct :)